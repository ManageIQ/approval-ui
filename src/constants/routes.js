const routes = {
  allrequests: {
    index: '/allrequests'
  },
  requests: {
    index: '/requests',
    addComment: '/requests/add-comment',
    approve: '/requests/approve',
    deny: '/requests/deny'
  },
  request: {
    index: '/request',
    addComment: '/request/add-comment',
    approve: '/request/approve',
    deny: '/request/deny'
  },
  workflows: {
    index: '/workflows',
    add: '/workflows/add-workflow',
    remove: '/workflows/remove',
    editInfo: '/workflows/edit-info',
    editGroups: '/workflows/edit-groups',
    editSequence: '/workflows/edit-sequence'
  }
};

export default routes;
