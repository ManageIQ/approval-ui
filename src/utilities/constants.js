import {
  CheckCircleIcon,
  InfoCircleIcon,
  ErrorCircleOIcon,
  ExclamationCircleIcon
} from '@patternfly/react-icons';
import React from 'react';
import requestsMessages from '../messages/requests.messages';

export const APPROVAL_API_BASE = `${process.env.BASE_PATH}/approval/v1.2`;
export const RBAC_API_BASE = `${process.env.BASE_PATH}/rbac/v1`;

export const decisionValues = {
  undecided: { displayName: requestsMessages.needsReview, color: 'blue', icon: <InfoCircleIcon /> },
  approved: { displayName: requestsMessages.approved, color: 'green', icon: <CheckCircleIcon /> },
  denied: { displayName: requestsMessages.denied, color: 'red', icon: <ExclamationCircleIcon /> },
  canceled: { displayName: requestsMessages.canceled, color: 'red', icon: <ErrorCircleOIcon /> },
  error: { displayName: requestsMessages.error, color: 'red', icon: <ExclamationCircleIcon /> }
};

// React intl does not support empty strings
export const untranslatedMessage = (defaultMessage = ' ') => ({ id: 'untranslated', defaultMessage });
export const APP_DISPLAY_NAME = {
  catalog: 'Automation Services Catalog',
  topology: 'Topological inventory'
};
