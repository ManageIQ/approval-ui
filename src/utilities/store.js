
import promiseMiddleware from 'redux-promise-middleware';
import ReducerRegistry, { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/files/ReducerRegistry';
import { notifications, notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';

import reduxLogger from 'redux-logger';
import thunk from 'redux-thunk';
import requestReducer, { requestsInitialState } from '../redux/reducers/request-reducer';
import workflowReducer, { workflowsInitialState } from '../redux/reducers/workflow-reducer';
import groupReducer, { groupsInitialState } from '../redux/reducers/group-reducer';
import rolesReducer, { rolesInitialState } from '../redux/reducers/roles-reducer';

const registry = new ReducerRegistry({}, [ thunk, promiseMiddleware(), notificationsMiddleware({
  errorTitleKey: [ 'message' ],
  errorDescriptionKey: [ 'errors', 'stack' ]
}), reduxLogger ]);

registry.register({
  requestReducer: applyReducerHash(requestReducer, requestsInitialState),
  workflowReducer: applyReducerHash(workflowReducer, workflowsInitialState),
  groupReducer: applyReducerHash(groupReducer, groupsInitialState),
  rolesReducer: applyReducerHash(rolesReducer, rolesInitialState),
  notifications
});

export default registry.getStore();
