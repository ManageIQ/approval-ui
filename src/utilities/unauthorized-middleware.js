import { approvalHistory } from '../router';

const unAuthorizedMiddleware = () => (dispatch) => (action) => {
  const nextAction = { ...action };
  if (action.type.match(/_REJECTED$/) && action?.payload?.redirect) {
    setTimeout(() => {
      approvalHistory.push(action.payload.redirect.pathname, {
        from: approvalHistory.location
      });
    });
    return;
  }

  return dispatch(nextAction);
};

export default unAuthorizedMiddleware;
