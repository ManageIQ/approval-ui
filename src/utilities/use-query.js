import { useLocation } from 'react-router-dom';

const useQuery = (requiredParams = []) => {
  const { search } = useLocation();
  const query = new URLSearchParams(search);
  return [
    requiredParams.reduce(
      (acc, curr) => ({
        ...acc,
        [curr]: query.get(curr)
      }),
      {}
    ),
    search,
    query
  ];
};

export default useQuery;
