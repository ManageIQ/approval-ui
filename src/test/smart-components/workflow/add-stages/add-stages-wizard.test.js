import React from 'react';
import thunk from 'redux-thunk';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store' ;
import { shallowToJson } from 'enzyme-to-json';
import AsyncSelect from 'react-select/async';

import { MemoryRouter, Route } from 'react-router-dom';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';
import { APPROVAL_API_BASE, RBAC_API_BASE } from '../../../../utilities/constants';
import AddWorkflow from '../../../../smart-components/workflow/add-stages/add-stages-wizard';
import { ADD_NOTIFICATION } from '@redhat-cloud-services/frontend-components-notifications/index';
import { mount } from 'enzyme/build/index';

describe('<AddWorkflow />', () => {
  let initialProps;
  let initialState;
  const middlewares = [ thunk, promiseMiddleware(), notificationsMiddleware() ];
  let mockStore;

  const ComponentWrapper = ({ store, children }) => (
    <Provider store={ store }>
      <MemoryRouter initialEntries={ [ '/workflows/', '/workflows/add-workflow/', '/workflows/' ] } initialIndex={ 1 }>
        { children }
      </MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    initialProps = {
      uuid: '123',
      postMethod: jest.fn(),
      handleChange: jest.fn()
    };

    initialState = {
      groupReducer: {
        groups: [{
          label: 'foo',
          value: 'bar'
        }]
      },
      workflowReducer: {
        workflows: { data: [{
          uuid: '123',
          name: 'SampleWorkflow'
        }]}
      }
    };
    mockStore = configureStore(middlewares);
  });

  it('should render correctly', () => {
    const store = mockStore(initialState);
    apiClientMock.get(`${APPROVAL_API_BASE}/workflows`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${RBAC_API_BASE}/groups`, mockOnce({ body: { data: []}}));
    const wrapper = shallow(<ComponentWrapper store={ store }><AddWorkflow { ...initialProps } /></ComponentWrapper>).dive();

    setImmediate(() => {
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  it('should post a warning message on Cancel', (done) => {
    const store = mockStore(initialState);

    apiClientMock.put(`${APPROVAL_API_BASE}/workflows/`, mockOnce((req, res) => {
      expect(req).toBeTruthy();
      return res.status(200);
    }));

    apiClientMock.get(`${APPROVAL_API_BASE}/workflows/`, mockOnce((req, res) => {
      expect(req).toBeTruthy();
      return res.status(200).body({ data: []});
    }));

    apiClientMock.get(`${RBAC_API_BASE}/groups/`, mockOnce((req, res) => {
      expect(req).toBeTruthy();
      return res.status(200).body({ data: []});
    }));

    const wrapper = mount(
      <ComponentWrapper store={ store }>
        <Route path="/workflows/add-workflow/" render={ () => <AddWorkflow { ...initialProps } /> } />
      </ComponentWrapper>
    );
    const expectedActions = expect.arrayContaining([
      expect.objectContaining({
        type: ADD_NOTIFICATION,
        payload: expect.objectContaining({
          title: 'Creating workflow',
          variant: 'warning',
          description: 'Creating workflow was cancelled by the user.',
          dismissable: true
        })
      }) ]);

    wrapper.find('Button').at(0).simulate('click');
    setImmediate(() => {
      expect(store.getActions()).toEqual(expectedActions);
      done();
    });
  });

  it('should travers over all wizard pages and call onSave function', async(done) => {
    expect.assertions(2);
    const store = mockStore(initialState);

    apiClientMock.get(`${APPROVAL_API_BASE}/templates`, mockOnce({ body: { data: [{ id: '123' }]}}));
    apiClientMock.post(`${APPROVAL_API_BASE}/templates/123/workflows`, mockOnce((req, res) => {
      expect(JSON.parse(req.body())).toEqual({ group_refs: [ '1' ]});
      done();
      return res.status(200);
    }));

    const wrapper = mount(
      <ComponentWrapper store={ store }>
        <Route path="/workflows/add-workflow/" render={ () => <AddWorkflow { ...initialProps } formData={ { wfGroups: [{}]} } /> } />
      </ComponentWrapper>
    );

    wrapper.find('button').at(1).simulate('click');
    wrapper.find('button.pf-c-button.pf-m-primary').first().simulate('click');
    setImmediate(async() => {
      wrapper.update();
      wrapper.find('button#add-workflow-stage').first().simulate('click');

      await act(async() => {
        wrapper.find(AsyncSelect).props().onChange({ value: '1', label: 'foo' });
      });
      wrapper.update();
      wrapper.find('button.pf-c-button.pf-m-primary').first().simulate('click');
      wrapper.update();

      /**
       * submit wizard
       */
      await act(async() => {
        wrapper.find('button.pf-c-button.pf-m-primary').first().simulate('click');
      });
    });
  });
});
