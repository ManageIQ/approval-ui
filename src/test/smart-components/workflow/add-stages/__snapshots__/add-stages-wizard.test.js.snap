// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddWorkflow /> should post a warning message on Cancel 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        "/workflows/",
        "/workflows/add-workflow/",
        "/workflows/",
      ]
    }
    initialIndex={1}
  >
    <AddWorkflow
      handleChange={[MockFunction]}
      initialValues={Object {}}
      postMethod={[MockFunction]}
      rbacGroups={Array []}
      uuid="123"
    />
  </MemoryRouter>
</ContextProvider>
`;
